Setup Rubocop and run with autocorrect

mokagio opened the pull request:
### Description
Sets up Rubocop for Ruby files linting and tracks the result of running it with autocorrect (`bundle exec rubocop -A`) through a number of key Ruby files.

There are issues that will need to be addressed manually, which I decided to leave for a later time. Ideally, the next time someone opens a Ruby file for editing, they can address those issues first.
 
### Testing instructions
N.A.

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary. — This is a non-user-facing change

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

mokagio committed:
Add Rubocop to the `Gemfile`

mokagio committed:
Re-bundle Ruby gems with latest Bundler (`2.3.6`)

mokagio committed:
Run `Gemfile` through Rubocop with autocorrect

mokagio committed:
Add `.rubocop.yml` configuration file

Lifted from WordPress iOS at:
https://github.com/wordpress-mobile/WordPress-iOS/blob/f1de7f600c5e14a6a1d92df81114facce379dc11/.rubocop.yml

mokagio committed:
Run `Podfile` through Rubocop with autocorrect

Notice that `Podfile.lock` changed, too, because of the checksum

mokagio committed:
Run Fastlane files through Rubocop with autocorrect

mokagio committed:
Run `replace_secrets.rb` build script through Rubocop w/ autocorrect

mokagio commented on .rubocop.yml:
```diff
@@ -0,0 +1,15 @@
+
+Naming/FileName:
+  Exclude:
+    - fastlane/Matchfile
```
I copied this from the WordPress repo.

mokagio commented on fastlane/Fastfile:
```diff
@@ -15,16 +17,16 @@ DERIVED_DATA_DIR = File.join(FASTLANE_DIR, 'DerivedData')
-MAIN_BUNDLE_IDENTIFIERS = %w[com.automattic.woocommerce]        # Registered in our main account, for development and AppStore
-ALPHA_BUNDLE_IDENTIFIERS = %w[com.automattic.alpha.woocommerce] # Registered in our Enterprise account, for AppCenter / Installable Builds
+MAIN_BUNDLE_IDENTIFIERS = %w[com.automattic.woocommerce].freeze        # Registered in our main account, for development and AppStore
+ALPHA_BUNDLE_IDENTIFIERS = %w[com.automattic.alpha.woocommerce].freeze # Registered in our Enterprise account, for AppCenter / Installable Builds
```
Actually, I think this change might back fire because of a Fastlane implementation detail that attempts to changed the array.

https://github.com/fastlane/fastlane/blob/b817148d4967ab6e38b8f1a8b81d1b3e60d6e645/match/lib/match/runner.rb#L84-L86](https://github.com/fastlane/fastlane/issues/11324

I'll need to remove it.

AliSoftware replied:
Nice catch!

mokagio committed:
Allow bundle identifier constant to be mutable because of Fastlane

See code comments for extra details

mokagio committed:
Run `Rakefile` through Rubocop with autocorrect

AliSoftware commented on Gemfile:
```diff
@@ -1,12 +1,16 @@
+gem 'rubocop', '~> 1.25'
+gem 'rubocop-rake', '~> 0.6'
 gem 'xcode-install'
 gem 'xcpretty-travis-formatter'
```
Not really related to this PR but this one could likely be removed 😜

AliSoftware commented on fastlane/Fastfile:
```diff
@@ -15,16 +17,20 @@ DERIVED_DATA_DIR = File.join(FASTLANE_DIR, 'DerivedData')
   'iPhone 8 Plus',
   'iPad Pro (12.9-inch) (2nd generation)',
   'iPad Pro (12.9-inch) (3rd generation)'
 ].freeze
```
Funny that Ruby seems to have a `frozen_string_literal: true` to auto-freeze string literals… but does not have the same for array literals like here… 🤷

AliSoftware commented on fastlane/Fastfile:
```diff
@@ -15,16 +17,20 @@ DERIVED_DATA_DIR = File.join(FASTLANE_DIR, 'DerivedData')
 ].freeze
 
+# rubocop:disable Style/MutableConstant
+# Fastlane mutates these internally, see
+# https://github.com/fastlane/fastlane/blob/b817148d4967ab6e38b8f1a8b81d1b3e60d6e645/match/lib/match/runner.rb#L84-L86
```
😢

AliSoftware commented on fastlane/Fastfile:
```diff
@@ -395,7 +403,7 @@ platform :ios do
 
     if options[:create_release]
-      archive_zip_path = File.dirname(Dir.pwd) + '/WooCommerce.xarchive.zip'
+      archive_zip_path = "#{File.dirname(Dir.pwd)}/WooCommerce.xarchive.zip"
```
Better fix:
```suggestion
      archive_zip_path = File.join(File.dirname(Dir.pwd), 'WooCommerce.xarchive.zip')
```

AliSoftware commented on fastlane/Fastfile:
```diff
@@ -422,7 +430,7 @@ platform :ios do
   #####################################################################################
   desc 'Builds and uploads an installable build'
-  lane :build_and_upload_installable_build do |options|
+  lane :build_and_upload_installable_build do |_options|
```
Alternatively, just remove the params pipes entirely… 🤷 
```suggestion
  lane :build_and_upload_installable_build do
```

AliSoftware commented on fastlane/Fastfile:
```diff
@@ -657,9 +665,9 @@ platform :ios do
-    })
+                                                           "workspace": '../WooCommerce.xcworkspace',
+                                                           "scheme": 'WooCommerceScreenshots'
+                                                         })
```
nit: maybe it would be more readable to go to the next line after `create(`? 🤷 

```ruby
    Snapshot.config = FastlaneCore::Configuration.create(
      Snapshot::Options.available_options,
      {
        "workspace": '../WooCommerce.xcworkspace',
        "scheme": 'WooCommerceScreenshots'
      }
    )
```

Also I wonder if Ruby isn't able to interpret a set of named parameters as a Hash by default when it's the last parameter? Which means that maybe it would even allow this?

```ruby
    Snapshot.config = FastlaneCore::Configuration.create(
      Snapshot::Options.available_options,
      workspace: '../WooCommerce.xcworkspace',
      scheme: 'WooCommerceScreenshots'
    )
```

mokagio replied:
I went for the first option, just to keep it "standard".

AliSoftware commented on fastlane/Fastfile:
```diff
@@ -820,33 +828,33 @@ platform :ios do
     UI.crash(e)
   ensure
-    file.close unless file.nil?
+    file&.close
```
Nice. But looking at the whole method, I wonder why we bothered creating this helper method while a one-liner ruby build-in like `File.write(path, contents)` would handle all of this (overwriting the previous file if it exists, creating a new one, handling potential exceptions, and ensuring closing the file in all cases)… 🤔 

(Probably for a separate PR anyway, but just wanted to drop it while I saw it)

AliSoftware commented on fastlane/Fastfile:
```diff
@@ -820,33 +828,33 @@ platform :ios do
 
-def simulator_version()
-  return '14.4'
+def simulator_version
+  '14.4'
```
Probably worth replacing this with a constant one day 😄

AliSoftware reviewed (approved):
Overall looks great 👍 

Left a couple of nits/suggestions for better manual fixes than the one made by autocorrect, so I disabled automerge and will let you address them before merging 🙂 

(Also left a couple of comments for improvements in the `Fastfile` but those are probably for a separate PR)

mokagio committed:
Merge remote-tracking branch 'origin/trunk' into add/rubocop

mokagio committed:
Merge remote-tracking branch 'origin/trunk' into add/rubocop

mokagio committed:
Remove `|_options|` parameter from lanes that don't use them

Rubocop added the `_` prefix, but we think it's better to remove the
whole parameter if unused.

See https://github.com/woocommerce/woocommerce-ios/pull/6118/files#r803897584

mokagio committed:
Remove two unused variables from `Fastfile`

mokagio committed:
Adjust `Snapshot.config` indentation

As per suggestion in PR.

See
https://github.com/woocommerce/woocommerce-ios/pull/6118/files#r803902759

mokagio committed:
Replace a path creation via interpolation with a `File.join` call

Co-authored-by: Olivier Halligon <olivier.halligon@automattic.com>

mokagio commented:
I addressed all the easy to address suggestions.

I'm then enable auto-merge.
